File: src/clojure_vulkan/math/glsl.clj
(ns clojure-vulkan.math.glsl
  (:refer-clojure :exclude [deftype float format])
  (:import (org.lwjgl.vulkan VK13)))

(defmacro deftype [type-name fields glsl-type-interface & methods]
  `(clojure.core/deftype ~type-name ~fields
     ~glsl-type-interface
     ~@(map (fn [[method-name argvec return]]
              (list method-name argvec
                    (if (list? return)
                      (apply * (map eval (rest return)))
                      (eval return)))) methods)))

(defprotocol GLSLType
  (components [this])
  (component-sizeof [this])
  (sizeof [this])
  (format [this]))

(deftype GLSLFloat []
  GLSLType
  (components [this] 1)
  (component-sizeof [this] Float/BYTES)
  (sizeof [this] (* 1 Float/BYTES))
  (format [this] VK13/VK_FORMAT_R32_SFLOAT))

(deftype GLSLVec2f []
  GLSLType
  (components [this] 2)
  (component-sizeof [this] Float/BYTES)
  (sizeof [this] (* 2 Float/BYTES))
  (format [this] VK13/VK_FORMAT_R32G32_SFLOAT))

(deftype GLSLVec3f []
  GLSLType
  (components [this] 3)
  (component-sizeof [this] Float/BYTES)
  (sizeof [this] (* 3 Float/BYTES))
  (format [this] VK13/VK_FORMAT_R32G32B32_SFLOAT))

(deftype GLSLVec4f []
  GLSLType
  (components [this] 4)
  (component-sizeof [this] Float/BYTES)
  (sizeof [this] (* 4 Float/BYTES))
  (format [this] VK13/VK_FORMAT_R32G32B32A32_SFLOAT))

(def float (GLSLFloat.))
(def vec2f (GLSLVec2f.))
(def vec3f (GLSLVec3f.))
(def vec4f (GLSLVec4f.))

(defn kw->type [kw]
  (case kw
    :float float
    :vec2 vec2f
    :vec3 vec3f
    :vec4 vec4f))

File: src/java/clojure_vulkan/math/GLSL.java
package clojure_vulkan.math;

import clojure.lang.Keyword;
import org.lwjgl.vulkan.VK13;

import java.util.HashMap;
import java.util.Map;

public class GLSL {
    private GLSL(){
    }
    public interface GLSLType {
        int componentsCount();

        int componentSizeof();

        int sizeof();

        int format();
    }

    private static class GLSLFloat implements GLSLType {
        public int componentsCount() {
            return 1;
        }

        public int componentSizeof() {
            return Float.BYTES;
        }

        public int sizeof() {
            return Float.BYTES;
        }

        public int format() {
            return VK13.VK_FORMAT_R32_SFLOAT;
        }
    }

    private static class GLSLVec2f implements GLSLType {
        public int componentsCount() {
            return 2;
        }

        public int componentSizeof() {
            return Float.BYTES;
        }

        public int sizeof() {
            return 2 * Float.BYTES;
        }

        public int format() {
            return VK13.VK_FORMAT_R32G32_SFLOAT;
        }
    }

    private static class GLSLVec3f implements GLSLType {
        public int componentsCount() {
            return 3;
        }

        public int componentSizeof() {
            return Float.BYTES;
        }

        public int sizeof() {
            return 3 * Float.BYTES;
        }

        public int format() {
            return VK13.VK_FORMAT_R32G32B32_SFLOAT;
        }
    }

    private static class GLSLVec4f implements GLSLType {
        public int componentsCount() {
            return 4;
        }

        public int componentSizeof() {
            return Float.BYTES;
        }

        public int sizeof() {
            return 4 * Float.BYTES;
        }

        public int format() {
            return VK13.VK_FORMAT_R32G32B32A32_SFLOAT;
        }
    }

    private static final Keyword
            FLOAT_KW = Keyword.intern("float"),
            VEC2_KW = Keyword.intern("vec2"),
            VEC3_KW = Keyword.intern("vec3"),
            VEC4_KW = Keyword.intern("vec4");

    public static final HashMap<Keyword, GLSLType> keywordToType = new HashMap<>(Map.of(
            FLOAT_KW, new GLSLFloat(),
            VEC2_KW, new GLSLVec2f(),
            VEC3_KW, new GLSLVec3f(),
            VEC4_KW, new GLSLVec4f()
    ));
}
